security/pkg/nodeagent/caclient/client.go

```diff
@@ -15,46 +15,75 @@
 package caclient
 
 import (
-	"crypto/x509"
-	"errors"
+	"encoding/base64"
 	"fmt"
 	"strings"
+	"time"
 
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
 
 	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/k8s/controller"
 	caClientInterface "istio.io/istio/security/pkg/nodeagent/caclient/interface"
+	citadel "istio.io/istio/security/pkg/nodeagent/caclient/providers/citadel"
 	gca "istio.io/istio/security/pkg/nodeagent/caclient/providers/google"
 )
 
-const googleCA = "GoogleCA"
+const (
+	googleCAName = "GoogleCA"
+	citadelName  = "Citadel"
+
+	retryInterval = time.Second * 2
+	maxRetries    = 100
+)
 
 // NewCAClient create an CA client.
 func NewCAClient(endpoint, CAProviderName string, tlsFlag bool) (caClientInterface.Client, error) {
-	var opts grpc.DialOption
-	if tlsFlag {
-		pool, err := x509.SystemCertPool()
+	switch CAProviderName {
+	case googleCAName:
+		return gca.NewGoogleCAClient(endpoint, tlsFlag)
+	case citadelName:
+		rootCert, err := getCATLSRootCertFromConfigMap()
 		if err != nil {
-			log.Errorf("could not get SystemCertPool: %v", err)
-			return nil, errors.New("could not get SystemCertPool")
+			return nil, err
```

Comment:
better log err before return

Reply:
The logging happens in main.go:
https://github.com/istio/istio/blob/release-1.1/security/cmd/node_agent_k8s/main.go#L66
So I think we can the defer error logging to upstream?

Reply:
sure, I missed the caller.